### DISCUSSION TITLE
Proposal 4 Addendum: Separating the intent of Layer 2 from the intentions of Layer 3

### DISCUSSION THREAD

**Discussion by mwherman2000** [trustoverip/trust-spanning-protocol#23] [Ideas]
While the multi-layered 5-category discussion of Layer 2 in Proposal 2 (https://github.com/trustoverip/trust-spanning-protocol/discussions/19) is valuable and stimulating, I believe in the spirit of, for example, the ISO networking stack (see below), it exceeds the intent or scope of a typical level in a network stack. I believe we need to undertake a better separation between:
1. the intent of the Layer 2 Trust Spanning Layer base protocol, from 
2. the intentions of the Layer 3 Trust Tasks super protocols.

**"I love you" Use Case**

With this understanding, if a Sender uses the Level 2 base protocol to (mistakenly) send an "I love you!" message to a large group of (unintended) Receivers, that mistake is entirely on the Sender for using the Layer 2 base protocol instead of a Layer 3 super protocol designed specifically for sending "I love you!" messages to one Receiver (and where the "single Receiver" limit is enforced as a business rule in the particular Layer 3 super protocol designed for that purpose).

**ISO Sam Stack Discussion**

Here's an enhanced visualization of 4 instances of an ISO networking protocol stack (aka 4 ISO Sam stacks).

Figure 1. ISO Sam Stacks
![SamSmith-IPv4-Spanning Layer-Routers-Annotated](https://user-images.githubusercontent.com/6101736/219681765-927f0f14-8974-4e35-9a40-d1e9b0de2fc2.png)

Don't get lost in the details of the above diagram. The purpose of this diagram is simply to remind us that:
> each layer in the ISO Sam stack model has a specific limited intent. 

A very narrow transformation of an outbound message/packet is defined at each ISO layer as it descends an Outbound Processing stack and a corresponding very narrow transformation of an inbound packet/message is defined at each ISO layer as it ascends through the Inbound Processing stack:

Figure 2. ISO Networking Model
![image](https://user-images.githubusercontent.com/6101736/219684539-20c91029-5813-4dcf-b2e0-cba8937580ee.png)

**Intent of the Layer 2 Trust Spanning Layer (Base) Protocol**

The (primary) intent of the Layer 2 Trust Spanning Layer base protocol should be (as a series of incrementally more detailed refinements):
1. To communicate
2. To communicate a message between a Sender and Receiver(s)
3. To communicate a message (and optional message attachments) between a Sender and Receiver(s)
4. To communicate, in a secure and private way, a message (and optional message attachments) between a Sender and Receiver(s)

Nothing more.  No "intentions" other than the above singular *intent*.  This is the essence of what is proposed in Proposal 3 (https://github.com/trustoverip/trust-spanning-protocol/discussions/21).  Proposal 3 is about the minimum required functionality (per the ISO model) expressed as the minimum number of Inbound and Outbound Processing steps to execute _a single message transfer between a single Sender and a single Receiver_.  The 4th definition of Layer 2 intent is sufficient to establish a basic level of _trust_. (Layer 3 is where higher levels of personal/business/romantic trust are established.)

NOTE: Support for multiple Receivers in this version of the model is achieved by looping over a collection of Receivers and treating each transmission as a single Sender-Receiver message transfer. More advanced semantics/scenrios such as 1 Sender : N Receivers to support scenarios like:

- Broadcasting a business Request For Quotation (RFQ) to a large group of supplier Receivers, or 
- Subscribers-Publishers scenario (N Subscribers : 1 Publisher) 

are the responsibilities of their respective Layer 3 Trust Tasks super protocols (see below).

UPDATE: Updated to show that some steps are optional depending on the requirements of a particular interaction/conversation.

Figure 3. Proposal 3 Web 7.0 TSLF Proposal as a Picture
![image](https://user-images.githubusercontent.com/6101736/219925262-6eecaabb-53cc-4074-bbfc-ce1fb534c560.png)

But how do we prevent Alice (Bob) from sending an "I love you!" message to every Bob (Alice) on the planet?  

First, if Alice or Bob choose to use the Layer 2 Trust Spanning Layer base protocol and mistakenly addresses the message to everyone in their respective organization, that's their mistake for using the Layer 2 Trust Spanning Layer base protocol.

How do we fix this?

**Intentions of the Layer 3 Trust Tasks (Super) Protocols**

Business intentions (as well as romantic intentions) are the responsibility of _super protocols_ in the Layer 3 Trust Tasks layer that are, in turn, defined on top of the Layer 2 Trust Spanning Layer base protocol (which is used as the underlying secure, private message transfer/transport layer).

In relative terms, the Layer 3 Trust Tasks super protocols are more _trusted_ relative to the Layer 2 Trust Spanning Layer base protocol because they are more business-context/application-context by design ...but they are still part of the Layer 3 Trust Tasks layer. This is because Layer 3 Trust Tasks super protocols are designed with more safeguards, and more business rules built into them (such as the "I love you!" limit of a single Receiver).  The following is a list of potential Layer 3 Trust Tasks super protocols (also from Daniel Hardman).

Figure 4. Layer 3 Trust Tasks super protocols (also from Daniel Hardman)
![image](https://user-images.githubusercontent.com/6101736/219694434-3cdae556-f383-4952-ad77-78f7ab494b83.png)

**Future Discussions**

A separate discussion (tabled for later) is the further use of a Layer 2 Trust Spanning Layer base protocol below Layer 2: between the Layer 2 Trust Spanning Layer and the Layer 1 Trust Foundation Layer. An example: using DIDCommm/HTTP as the basis for a DID Resolution protocol solution - so that an entire decentralized systems platform can be enabled using a single Trust Spanning Layer base protocol.  Here's a diagram (a Sam stack) to stimulate this follow-on conversation.

Figure 5. Using DIDCommm/HTTP as the basis for a DID Resolution
![image](https://user-images.githubusercontent.com/6101736/219696693-abc2b173-8320-4648-ac95-885151331c5b.png)

The End (for today ...at least for now)

**Comment by dhh1128** [comment]
Michael, I am not and never have been concerned about Alice mistakenly sending an "I love you" to the wrong audience. I have been concerned about Alice composing her message to one audience, and having it delivered by a malicious intermediary to the wrong audience. If Alice herself addresses or delivers her message to the wrong audience, that is her problem, not the TSP's.

Although a specialized layer 3 protocol could be written to send "I love you" messages, I think that misses my point. No matter what kind of message we're talking about, the same intent issues arise: who is the audience, how does this message relate to other things that audience already knows about, etc. I claim that these issues must be defined for ALL messages that flow over TSP. That is not because every message has a constrained audience or depends on other messages -- many don't -- but rather because I need to be able to tell the difference between these two states in ALL cases, if I am going to be able to reason about the trust properties that apply to an arbitrary mix of context created by composable trust tasks sitting above the TSP.

**Reply by mwherman2000** [reply]
When you're talking about _intent_ and _context_, my mind says the answer is simple: _it's whatever the state is of the (corresponding) current conversation is_. 

In a business scenario (e.g. a product or services procurement scenario), the intent and context of an inbound message is determined by the current state of the particular instance of the particular business process (a correlated, executing instance of a formally defined business process workflow template).  This is still Layer 3.  This does however say that a Layer 3 super protocol is now defined as something broader and more specific.  A Layer 3 super protocol includes (among other things):
1. Process-specific messages formats/schemas
2. Process-specific (or generic standardized) message attachments schemas
3. A specific executable business process workflow template (e.g an OASIS BPMN process workflow template serialized in the OASIS BPMN defined XML serialization format)
4. Additional governance such as business rules, policies, procedures, process (3P) documentation

Assumption: the Receiver and Sender agents each have an embedded or decentralized workflow engine available to run the business process workflows.  This is true in Web 7.0.

Given the above, I'm not certain I'm understanding your comments.  ...but I still think the above is my answer. :-)

**Reply by mwherman2000** [reply]
> I have been concerned about Alice composing her message to one audience, and having it delivered by a malicious intermediary to the wrong audience.

How is this possible @dhh1128? 

If Alice sends, for example, a DIDComm Message addressed to Bob (auth-encrypted so that only Bob can read it) and a malicious party intercepts the auth-encrypted message itself and/or redirects it to Charlie, what does it matter? Neither the interloper nor Charlie *cannot* decrypt and read the message, n'est pas?

**Reply by dhh1128** [reply]
There is no authcrypt in Sam's proposal, or in mine.

There *is* an additional encryption layer, but it is optional and above the base. So we can't assume that encryption will protect us.

**Reply by mwherman2000** [reply]
Now we've gotten to the root of the problem. :-) If we try to find some common ground, one could say steps S6 and R2 in https://user-images.githubusercontent.com/6101736/219690953-bce7d2c4-79a1-4cc2-860e-2ee323222619.png are optional and handled in a different layer but this is sort of just semantics in the sense of how you choose to factor the layers (but I guess this is the essence of what we're discussing).

**Reply by dhh1128** [reply]
This is actually not much different from DIDComm; it, too, makes encryption optional.

**Reply by dhh1128** [reply]
>When you're talking about intent and context, my mind says the answer is simple: it's whatever the state is of the (corresponding) current conversation is.

Sam's proposal for the TSP is stateless. It is JUST an authentic message, in isolation. You might impute state to the messages as you arrive, but you have no justification for believing that they arrive in the order they were sent.

My proposal is for the TSP to *define* state but not require it. Sometimes interactions ARE stateless, so we don't want to constantly require state management when it's useless. But I want to be able to tell the difference between a message that is intended to be stateless, and a message that is intended to be evaluated in context. That's why I want to define state in TSP.

**Reply by mwherman2000** [reply]
> My proposal is for the TSP to _define_ state but not require it. Sometimes interactions ARE stateless, so we don't want to constantly require state management when it's useless.

@dhh1128 In the section The Intent of Layer 2..., all communications are in fact stateless (maybe unstateful is a more precise term) as they should be. 

It is the responsibility of the Receiver and Sender to maintain what they think is the current state of their side of the conversation ...the same as in real-life business interactions. For example if  I sent you the message "too bad you're going to be on vacation in April", it has intent based on the context and current state of a separate conversation you and I started this morning.

The Sender and Receiver don't even need to be executing the same business process workflow; for example, two companies using different ERP systems with noticeably different executable procurement business process workflow templates will each maintain their private state of their private business processes. The Sender's state may in fact be unintelligible to the Receiver and visa versa - if the state was to be exchanged.

**Comment by mwherman2000** [comment]
> who is the audience, how does this message relate to other things that audience already knows about, etc. I claim that these issues must be defined for ALL messages that flow over TSP. That is not because every message has a constrained audience or depends on other messages -- many don't

Although it's considered a dirty word, there could be some provision in the Layer 2 for some sort of simple correlation token identifier (e.g. a random DID) that can be used to group collections of messages into a conversation - the same way that SMTP email message conversations (threads) are correlated (https://stackoverflow.com/questions/288757/how-to-identify-email-belongs-to-existing-thread-or-conversation).

...else leave it to each Layer 3 super protocol to define their own correlation identifiers.

In a formal business process scenario (e.g. procurement scenario), the correlation identifier could, for example, be the DID that identifies the specific instance (maybe even specific instance + current state) of the executing business workflow template.

**Reply by dhh1128** [reply]
Exactly. That's why I proposed a "thread ID" at that layer, to do exactly that thing.

**Comment by mwherman2000** [comment]
@dhh1128 So... are we agreeing on everything? I've lost track of if/where we differ. :-)  

I still prefer the Proposal 3 diagram. :-) ...maybe it represents the next level of refinement of the Proposal 2 diagram.

**Reply by dhh1128** [reply]
I think we are aligned on a lot of stuff. But there are still some differences. For example, you say that the intent is to communicate in a secure and private way. I think privacy may not be the intent. If the press office of a company or government official wants to tweet something, they don't care about privacy. But they certainly care about trust. So privacy sits above the base.

I also think "secure" in your phrase might mean "confidential" -- and if it is, that's also a separate layer. If, on the other hand, you mean "authentic" instead of "confidential" or instead of "authentic AND confidential," then we're pretty close to agreement. :-)

**Comment by dhh1128** [comment]
We're talking about two different types of state, I think.

You are of course right that each party experiences their own state that
may be meaningful to no other party on the planet. This is always true. But
what I am talking about is the state of a formally defined and shared
interaction: do both parties perceive that three messages have occurred in
this (chat | chess game | payment | loan application)? It is normal for
these perceptions to differ; every time Alice sends but Bob has not yet
perceived, she will have a different perception of the state of the
interaction than he will. If the interaction is supposed to be stateful,
then the parties care about this discrepancy and attempt to reason about it
and reconcile it, because their goals depend upon it. A chess game is like
this. On the other hand, if they have no interest in achieving some degree
of alignment about what has happened in an interaction, then it is
stateless. A temperature sensor that broadcasts readings to the public is
stateless.

**Reply by mwherman2000** [reply]
> But what I am talking about is the state of a formally defined and shared
interaction: do both parties perceive that three messages have occurred in
this (chat | chess game | payment | loan application)? It is normal for
these perceptions to differ; every time Alice sends but Bob has not yet
perceived, she will have a different perception of the state of the
interaction than he will.

@dhh1128 Absolutely. It's worse than that: the intent and state of the other party in a particular conversation are indeterminant - even in a particular formal process. 

This is certainly true in _chat_ (see example scenario below).  It is even true in loan origination and employee/contractor recruiting these days with human-mediated bots taking over these responsibilities.  You _may_ think everything is going grand but you have no idea to know what the loan officer or recruiter is actually thinking (especially no voice intonation or body language cues). 

In the latter case, the recruiter may have already chosen the successful candidate but might be formally/legally required to interview N people before formalizing her/his decision.  

Same with the loan officer in the former case, he/she may have a quota of applications to consider (as a metric) and/or the loan company may have a quota for loan applications with applicants from a particular region or demographic background.

Here's a chat scenario:

1. Sender Alice, in her mind, invokes the "I strongly believe in X and want to convince Bob that he should also believe in X as I do" _personal process_ and begins sending messages to Bob about why X is good.
2. Receiver Bob doesn't really care about X and if he did, he would strongly disagree with Alice about X but he knows Alice is extremely bright. So Bob engages using his "I don't really care about X but let's see what Alice has to say" _personal process_ and begins replying to Alice with what are, in effect, meaningless, inconsequential acknowledgments.
3. This conversation between Alice and Bob about X continues for hours, days, weeks, months, or years.

**Reply by dhh1128** [reply]
@mwherman2000 : you are correct, but I think you are pushing so hard for indeterminism that you are omitting something important: _The purpose of protocols is to provide rules that let other parties reason in predictable (albeit limited) ways about state._ So, while shared state is ultimately indeterminate, it is not unconstrained. That distinction matters a lot.

Without a doubt, Alice never knows what Bob is thinking, and she also cannot guarantee, *ever*, that she knows Bob's own perspective of his state. However, if Alice and Bob are playing a chess game, she can predict the possible states that Bob might have for his view of the chess board, and she can follow orderly rules to reason about state evolution. Why he's playing chess is out of scope. Whether he's bored is out of scope. Whether he will send her a non-sequitur or an invalid repeat of his last move is unknowable. But since the game uses a turn-taking protocol, Alice may know with perfect confidence that, ACCORDING TO THE RULES OF THE PROTOCOL, Bob's view of the state of the chess game can differ from her own by at most one turn. At time T, Alice knows that the game transcript included 7 moves, with Bob's being the 7th. Thus it is her turn to make the 8th move. She creates and transmits a message expressing her choice for move 8. It is possible that while she does this, Bob is sending a "cancel the game" message (and has already ended the game, in his mind). It is possible that Bob just got hit by a meteor. It is possible that Bob never receives her 8th move. But it is not possible that Bob thinks the protocol allows him to claim that only 6 moves have elapsed, if he transmitted a signed version of the message where he made move 7. It is also not possible that Bob thinks he can make move 9, if Alice hasn't sent move 8 yet. *The protocol says what reasoning can be done about state.*

Different protocols have different state evolution rules. Chess involves turn-taking. Online auctions definitely don't. Chat does not, either (not necessarily, anyway). Therefore, what constitutes "state" for chat will be different. But it won't be nothing. Although Alice can never reason reliably about the relative order of a pair of messages between her and Bob that pass each other as they cross they wire, she can know the relative order of her own transmissions, and if need be, she can embed in one of her messages a proof of existence of Bob's message to guarantee that he knows hers came after his (was a reply). Etc.

**Reply by mwherman2000** [reply]
@dhh1128 Chess is not a good example because the _shared state_ of the chess board (real or virtual) is transparent to both parties all of the time. 
The only personal state that each of Alice and Bob have is the current context of their internal strategy (i.e. each has their own private internal chess "business process" executing).

**Reply by dhh1128** [reply]
It may be that we are saying the same thing, but are using words that make us feel like we are disagreeing.

I claim that chess is exactly like many other DIDComm protocols insofar as its statefulness is concerned. Alice and Bob are not guaranteed to have a transparent or common view of the chess board. In 100% of all chess games, there will regularly be times between Alice's transmission of a move and Bob's reception of it, when she and Bob will see the board differently. Also, either one could have an error that the other doesn't know about yet. And Bob can abandon the game and never tell Alice. Etc. There is no central or shared location that functions as a source of truth -- only each party's record of what they believe has happened.

Separate from the inconsistent-but-contrained-by-rules-of-the-protocol perspective that Alice and Bob each have about the chessboard, Alice and Bob *also* each have private thoughts about strategy. Alice may be executing a queen's gambit, and Bob may be oblivious. Bob may be losing the chess game in purpose to ingratiate himself with Alice. These are the things I think about when I see your phrase "internal strategy..their own internal chess 'business process'). Am I mapping between your thoughts and mine accurately?

**Comment by mwherman2000** [comment]
@dhh1128 I've updated _Figure 3. Proposal 3 Web 7.0 TSLF Proposal as a Picture_ in https://github.com/trustoverip/trust-spanning-protocol/discussions/23#discussion-4869614 to illustrate which Outbound and Inbound Message Processing steps that are optional in the Web 7.0 model ...the minimum being 3 steps outbound and 2 steps inbound.  Here's an illustration of this extreme scenario.  

Figure 6. Degenerate, Minimal Outbound and Inbound Message Processing Steps
![image](https://user-images.githubusercontent.com/6101736/219925411-a29d51ed-c9a7-45f4-a40f-32efce1c1f0a.png)

NOTE: I've also added step _S4. Add Correlation Identifier (co-id)_ and corresponding reference to the co-id in step R6 in _Figure 3. Proposal 3 Web 7.0 TSLF Proposal as a Picture_ (https://github.com/trustoverip/trust-spanning-protocol/discussions/23#discussion-4869614).

**Comment by mwherman2000** [comment]
Good footnote to this conversation that appears to be dying on the vine: https://twitter.com/mwherman2000/status/1628042135378223105?t=iZMtKENRoRwhU8IYFJ6UHA&s=19

**Comment by mwherman2000** [comment]
@dhh1128 @SmithSamuelM side note: I know you and Sam are pushing hard to KERIze the ToIP Trust Spanning Layer model but I want to drop a word of caution here: don't be so over zealous that we end up with what the JSON-LD/RDF Block of partners did to the W3C VC2 data model spec: made only useable if you force yourself to use @context and @vocab to serve no useful purpose other than serve as useless Christmas tree decorations.
Allow for a truly layered model that developers will understand...like the one I've proposed above

**Comment by mwherman2000** [comment]
The formal summary presentation can be found here: https://github.com/trustoverip/trust-spanning-protocol/discussions/27
